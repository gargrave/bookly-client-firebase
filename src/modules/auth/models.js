// @flow
import type {
  LoginErrors,
  LoginUser,
  PasswordReset,
  PasswordResetErrors,
  RegisterErrors,
  RegisterUser,
  User,
  UserWithMetadata,
} from './flowtypes';

export const userModel = {
  empty(): User {
    return {
      displayName: '',
      email: '',
      emailVerified: false,
      id: '',
      lastLogin: '',
      registered: '',
    };
  },

  fromAPI(user: UserWithMetadata): User {
    const {
      displayName,
      email,
      emailVerified,
      uid,
    } = user;
    const {
      creationTime,
      lastSignInTime,
    } = user.metadata;

    return {
      displayName,
      email,
      emailVerified,
      id: uid,
      lastLogin: lastSignInTime,
      registered: creationTime,
    };
  },
};

export const loginUserModel = {
  empty(): LoginUser {
    return {
      email: '',
      password: '',
    };
  },

  emptyErrors(): LoginErrors {
    return {
      email: '',
      password: '',
    };
  },

  toAPI(data: LoginUser): LoginUser {
    let payload: LoginUser = {
      email: data.email,
      password: data.password,
    };
    return payload;
  },
};

export const registerUserModel = {
  empty(): RegisterUser {
    return {
      email: '',
      password: '',
      passwordConfirm: '',
    };
  },

  emptyErrors(): RegisterErrors {
    return {
      email: '',
      password: '',
      passwordConfirm: '',
    };
  },

  toAPI(data: RegisterUser): RegisterUser {
    return {
      email: data.email,
      password: data.password,
      passwordConfirm: data.passwordConfirm,
    };
  },
};

export const passwordResetModel = {
  empty(): PasswordReset {
    return {
      email: '',
    };
  },

  emptyErrors(): PasswordResetErrors {
    return {
      email: '',
    };
  },
};
