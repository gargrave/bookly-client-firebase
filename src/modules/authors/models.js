// @flow
import type { Author, AuthorErrors } from './flowtypes';
import type { FbDoc } from '../core/flowtypes';

export const authorModel = {
  empty(): Author {
    return {
      firstName: '',
      lastName: '',
    };
  },

  emptyErrors(): AuthorErrors {
    return {
      firstName: '',
      lastName: '',
    };
  },

  editable(author: Author): Author {
    return {
      created: author.created,
      firstName: author.firstName,
      id: author.id,
      lastName: author.lastName,
    };
  },

  toAPI(data: Author): Author {
    let payload: any = {
      firstName: data.firstName.trim() || '',
      lastName: data.lastName.trim() || '',
    };

    ['id', 'created'].forEach((val) => {
      if (data[val]) {
        payload[val] = data[val];
      }
    });

    return payload;
  },

  fromAPI(doc: FbDoc): Author {
    const {
      firstName,
      lastName,
      created,
      updated,
    } = doc.data();

    return {
      id: doc.id,
      firstName,
      lastName,
      created,
      updated,
    };
  },
};
