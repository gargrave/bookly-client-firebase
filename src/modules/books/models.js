// @flow
import type { Author } from '../authors/flowtypes';
import type { FbDoc } from '../core/flowtypes';
import type { Book, BookErrors } from './flowtypes';

function hydrateAuthor(authors, id): Author {
  const author = authors.find((a) => a.id === id);
  return author || {
    firstName: '',
    lastName: '',
  };
}

export function refreshBookAuthor(book: Book, authors: Author[]) {
  book.author = hydrateAuthor(authors, book.author.id || book.authorId);
}

export const bookModel = {
  empty(): Book {
    return {
      author: {
        id: '',
        firstName: '',
        lastName: '',
      },
      title: '',
    };
  },

  emptyErrors(): BookErrors {
    return {
      title: '',
      author: '',
    };
  },

  editable(book: Book): Book {
    return {
      author: book.author,
      created: book.created,
      id: book.id,
      title: book.title.trim(),
    };
  },

  toAPI(data: Book): Book {
    let payload: any = {
      title: data.title.trim(),
      authorId: data.author.id,
    };

    ['id', 'created'].forEach((val) => {
      if (data[val]) {
        payload[val] = data[val];
      }
    });

    return payload;
  },

  fromAPI(doc: FbDoc, authors: Author[]): Book {
    const {
      title,
      authorId,
      created,
      updated,
    } = doc.data();

    return {
      id: doc.id,
      title,
      author: hydrateAuthor(authors, authorId),
      created,
      updated,
    };
  },
};
