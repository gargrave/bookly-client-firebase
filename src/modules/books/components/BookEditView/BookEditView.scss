@import '../../../../styles/_views';

.bookEditView {
  @include viewWrapper;
}
