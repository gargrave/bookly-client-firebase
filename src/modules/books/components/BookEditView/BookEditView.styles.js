import styled from 'react-emotion'

import { views } from '../../../../styles'

export const BookEditView = styled('div')`
  ${views.viewWrapper};
`
