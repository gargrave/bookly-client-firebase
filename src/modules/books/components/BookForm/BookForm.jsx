// @flow
import React from 'react'
import { array, bool, func, object, shape, string } from 'prop-types'

import type { Author } from '../../../authors/flowtypes'
import type { Book, BookErrors } from '../../flowtypes'

import AuthorSelect from '../../../authors/components/AuthorSelect'
import Form from '../../../common/components/Form/Form'
import InputField from '../../../common/components/InputField/InputField'

type Props = {
  authors: Author[],
  book: Book,
  disabled?: boolean,
  errors: BookErrors,
  onAuthorChange: Function,
  onCancel: Function,
  onInputChange: Function,
  onSubmit: Function,
  preselectedAuthor?: Author,
  submitDisabled?: boolean,
  topLevelError?: string,
}

const BookForm = ({
  authors,
  book,
  disabled,
  errors,
  onAuthorChange,
  onCancel,
  onInputChange,
  onSubmit,
  preselectedAuthor,
  submitDisabled = false,
  topLevelError,
}: Props) => {
  const { author, sortBy, title } = book

  return (
    <Form
      classes={['book-form']}
      disabled={disabled}
      onCancel={onCancel}
      onSubmit={onSubmit}
      submitDisabled={submitDisabled}
      topLevelError={topLevelError}
    >
      <InputField
        boundValue={title}
        disabled={disabled || false}
        error={errors.title}
        label="Title"
        name="title"
        onInputChange={onInputChange}
      />

      <InputField
        boundValue={sortBy}
        disabled={disabled || false}
        error={errors.sortBy}
        label="Sort By"
        name="sortBy"
        onInputChange={onInputChange}
        placeholder={title}
      />

      <AuthorSelect
        author={preselectedAuthor || author}
        authors={authors}
        disabled={disabled || false}
        error={errors.author}
        onChange={onAuthorChange}
      />
    </Form>
  )
}

BookForm.propTypes = {
  authors: array.isRequired,
  book: shape({
    title: string.isRequired,
    author: object.isRequired,
  }).isRequired,
  disabled: bool,
  errors: shape({
    title: string.isRequired,
    author: string.isRequired,
  }).isRequired,
  onAuthorChange: func.isRequired,
  onCancel: func.isRequired,
  onInputChange: func.isRequired,
  onSubmit: func.isRequired,
  preselectedAuthor: object,
  submitDisabled: bool,
  topLevelError: string,
}

export default BookForm
