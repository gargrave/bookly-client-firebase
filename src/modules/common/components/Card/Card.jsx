// @flow
import React from 'react'
import { any, bool, func } from 'prop-types'
import styled from 'react-emotion'

import { buildClass } from '../../../../utils/cssHelpers'

import CardHeader from './CardHeader'
import CardSpacer from './CardSpacer'
import CardTextLine from './CardTextLine'
import CardTextList from './CardTextList'

import { colors, shadows } from '../../../../styles/'

type Props = {
  children?: any,
  hoverable?: boolean,
  onClick?: Function,
}

const Styled = styled('div')`
  ${shadows.medium};

  background-color: ${colors.cardBg};
  border: solid 1px ${colors.cardBorder};
  border-radius: 5px;
  margin: 8px auto;
  max-width: 550px;
  padding: 10px 20px;
  text-align: center;

  &.hoverable:hover {
    background: ${colors.cardHoveredBg};
    cursor: pointer;
  }
`

const Card = ({ children, hoverable = false, onClick }: Props) => {
  return (
    <Styled className={buildClass({ hoverable })} onClick={onClick}>
      {children}
    </Styled>
  )
}

Card.Header = CardHeader
Card.Spacer = CardSpacer
Card.TextLine = CardTextLine
Card.TextList = CardTextList

Card.propTypes = {
  children: any,
  hoverable: bool,
  onClick: func,
}

export default Card
