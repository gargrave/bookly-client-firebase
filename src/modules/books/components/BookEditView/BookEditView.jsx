// @flow
import React from 'react'
import { array, bool, func, object, string } from 'prop-types'

import type { Author } from '../../../authors/flowtypes'
import type { Book, BookErrors } from '../../flowtypes'

import BookForm from '../BookForm/BookForm'
import Card from '../../../common/components/Card/Card'

import * as S from './BookEditView.styles'

type Props = {
  authors: Author[],
  book: Book,
  disabled: boolean,
  errors: BookErrors,
  onAuthorChange: Function,
  onCancel: Function,
  onInputChange: Function,
  onSubmit: Function,
  submitDisabled?: boolean,
  topLevelError?: string,
}

const BookEditView = ({
  authors,
  book,
  disabled,
  errors,
  onAuthorChange,
  onCancel,
  onInputChange,
  onSubmit,
  submitDisabled = false,
  topLevelError,
}: Props) => {
  return (
    <S.BookEditView>
      <Card>
        <Card.Header text="Update Book" />
        <BookForm
          authors={authors}
          book={book}
          disabled={disabled}
          errors={errors}
          onAuthorChange={onAuthorChange}
          onCancel={onCancel}
          onInputChange={onInputChange}
          onSubmit={onSubmit}
          submitDisabled={submitDisabled}
          topLevelError={topLevelError}
        />
      </Card>
    </S.BookEditView>
  )
}

BookEditView.propTypes = {
  authors: array.isRequired,
  book: object.isRequired,
  disabled: bool,
  errors: object.isRequired,
  onAuthorChange: func.isRequired,
  onCancel: func.isRequired,
  onInputChange: func.isRequired,
  onSubmit: func.isRequired,
  submitDisabled: bool,
  topLevelError: string,
}

export default BookEditView
