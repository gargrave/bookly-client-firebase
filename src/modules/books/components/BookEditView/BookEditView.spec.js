import React from 'react';
import { shallow } from 'enzyme';

import { authorMocks, bookMocks } from '../../../../globals/mocks/';
import { bookModel } from '../../models';

import BookForm from '../BookForm/BookForm';
import Card from '../../../common/components/Card/Card';

import BookEditView from './BookEditView';

const testBook = bookMocks[0];

const defaultProps = Object.freeze({
  authors: authorMocks,
  book: testBook,
  disabled: false,
  errors: bookModel.emptyErrors(),
  onAuthorChange: jest.fn(),
  onCancel: jest.fn(),
  onInputChange: jest.fn(),
  onSubmit: jest.fn(),
  submitDisabled: false,
  topLevelError: '',
});

function getComponent(extraProps = {}) {
  const props = Object.assign({}, defaultProps, extraProps);
  return shallow(<BookEditView {...props} />);
}

describe('BookEditView', () => {
  let component;

  test('matches the snapshot', () => {
    component = getComponent();
    expect(component).toMatchSnapshot();
  });

  test('renders correctly', () => {
    component = getComponent();
    const parentClass = '.bookly-book-edit-view';
    expect(component.find(parentClass)).toHaveLength(1);
  });

  test('renders one BookForm component', () => {
    component = getComponent();
    expect(component.find(BookForm)).toHaveLength(1);
  });

  test('renders one Card component', () => {
    component = getComponent();
    expect(component.find(Card)).toHaveLength(1);
  });
});
