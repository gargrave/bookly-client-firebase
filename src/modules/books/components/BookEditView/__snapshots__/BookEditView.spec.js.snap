// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BookEditView matches the snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <BookEditView
    authors={
      Array [
        Object {
          "bookCount": 1,
          "created": 2017-07-01T01:16:43.881Z,
          "firstName": "Aldous",
          "id": "oiw223h3qndlksadffl3ksdf",
          "lastName": "Huxley",
          "updated": 2017-07-01T01:16:43.881Z,
        },
        Object {
          "bookCount": 1,
          "created": 2017-07-06T22:02:11.504Z,
          "firstName": "Kurt",
          "id": "9328fhs938hfdsdblkhf98",
          "lastName": "Vonnegut",
          "updated": 2017-07-06T22:02:11.504Z,
        },
        Object {
          "bookCount": 1,
          "created": 2017-08-17T20:47:35.232Z,
          "firstName": "George R. R.",
          "id": "92nd9ab3kjasegfhasf783fu",
          "lastName": "Martin",
          "updated": 2017-08-17T20:47:35.232Z,
        },
        Object {
          "bookCount": 1,
          "created": 2017-07-06T22:02:17.051Z,
          "firstName": "Tom",
          "id": "98fn98w3ucbe8c7b276v",
          "lastName": "Robbins",
          "updated": 2017-08-17T21:27:35.885Z,
        },
        Object {
          "bookCount": 2,
          "created": 2017-07-06T22:02:17.051Z,
          "firstName": "Stephen",
          "id": "93ucbe8c8fn98w7b276v",
          "lastName": "King",
          "updated": 2017-08-17T21:27:35.885Z,
        },
      ]
    }
    book={
      Object {
        "author": Object {
          "bookCount": 1,
          "created": 2017-07-01T01:16:43.881Z,
          "firstName": "Aldous",
          "id": "oiw223h3qndlksadffl3ksdf",
          "lastName": "Huxley",
          "updated": 2017-07-01T01:16:43.881Z,
        },
        "created": 2018-01-14T01:36:26.220Z,
        "id": "sv3hmbefpc95MRXQsTHr",
        "sortBy": "",
        "title": "Brave New World",
        "updated": 2018-01-14T01:36:26.220Z,
      }
    }
    disabled={false}
    errors={
      Object {
        "author": "",
        "sortBy": "",
        "title": "",
      }
    }
    onAuthorChange={[Function]}
    onCancel={[Function]}
    onInputChange={[Function]}
    onSubmit={[Function]}
    submitDisabled={false}
    topLevelError=""
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": <Card>
        <CardHeader
          text="Update Book"
        />
        <BookForm
          authors={
            Array [
              Object {
                "bookCount": 1,
                "created": 2017-07-01T01:16:43.881Z,
                "firstName": "Aldous",
                "id": "oiw223h3qndlksadffl3ksdf",
                "lastName": "Huxley",
                "updated": 2017-07-01T01:16:43.881Z,
              },
              Object {
                "bookCount": 1,
                "created": 2017-07-06T22:02:11.504Z,
                "firstName": "Kurt",
                "id": "9328fhs938hfdsdblkhf98",
                "lastName": "Vonnegut",
                "updated": 2017-07-06T22:02:11.504Z,
              },
              Object {
                "bookCount": 1,
                "created": 2017-08-17T20:47:35.232Z,
                "firstName": "George R. R.",
                "id": "92nd9ab3kjasegfhasf783fu",
                "lastName": "Martin",
                "updated": 2017-08-17T20:47:35.232Z,
              },
              Object {
                "bookCount": 1,
                "created": 2017-07-06T22:02:17.051Z,
                "firstName": "Tom",
                "id": "98fn98w3ucbe8c7b276v",
                "lastName": "Robbins",
                "updated": 2017-08-17T21:27:35.885Z,
              },
              Object {
                "bookCount": 2,
                "created": 2017-07-06T22:02:17.051Z,
                "firstName": "Stephen",
                "id": "93ucbe8c8fn98w7b276v",
                "lastName": "King",
                "updated": 2017-08-17T21:27:35.885Z,
              },
            ]
          }
          book={
            Object {
              "author": Object {
                "bookCount": 1,
                "created": 2017-07-01T01:16:43.881Z,
                "firstName": "Aldous",
                "id": "oiw223h3qndlksadffl3ksdf",
                "lastName": "Huxley",
                "updated": 2017-07-01T01:16:43.881Z,
              },
              "created": 2018-01-14T01:36:26.220Z,
              "id": "sv3hmbefpc95MRXQsTHr",
              "sortBy": "",
              "title": "Brave New World",
              "updated": 2018-01-14T01:36:26.220Z,
            }
          }
          disabled={false}
          errors={
            Object {
              "author": "",
              "sortBy": "",
              "title": "",
            }
          }
          onAuthorChange={[Function]}
          onCancel={[Function]}
          onInputChange={[Function]}
          onSubmit={[Function]}
          submitDisabled={false}
          topLevelError=""
        />
      </Card>,
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <CardHeader
            text="Update Book"
          />,
          <BookForm
            authors={
              Array [
                Object {
                  "bookCount": 1,
                  "created": 2017-07-01T01:16:43.881Z,
                  "firstName": "Aldous",
                  "id": "oiw223h3qndlksadffl3ksdf",
                  "lastName": "Huxley",
                  "updated": 2017-07-01T01:16:43.881Z,
                },
                Object {
                  "bookCount": 1,
                  "created": 2017-07-06T22:02:11.504Z,
                  "firstName": "Kurt",
                  "id": "9328fhs938hfdsdblkhf98",
                  "lastName": "Vonnegut",
                  "updated": 2017-07-06T22:02:11.504Z,
                },
                Object {
                  "bookCount": 1,
                  "created": 2017-08-17T20:47:35.232Z,
                  "firstName": "George R. R.",
                  "id": "92nd9ab3kjasegfhasf783fu",
                  "lastName": "Martin",
                  "updated": 2017-08-17T20:47:35.232Z,
                },
                Object {
                  "bookCount": 1,
                  "created": 2017-07-06T22:02:17.051Z,
                  "firstName": "Tom",
                  "id": "98fn98w3ucbe8c7b276v",
                  "lastName": "Robbins",
                  "updated": 2017-08-17T21:27:35.885Z,
                },
                Object {
                  "bookCount": 2,
                  "created": 2017-07-06T22:02:17.051Z,
                  "firstName": "Stephen",
                  "id": "93ucbe8c8fn98w7b276v",
                  "lastName": "King",
                  "updated": 2017-08-17T21:27:35.885Z,
                },
              ]
            }
            book={
              Object {
                "author": Object {
                  "bookCount": 1,
                  "created": 2017-07-01T01:16:43.881Z,
                  "firstName": "Aldous",
                  "id": "oiw223h3qndlksadffl3ksdf",
                  "lastName": "Huxley",
                  "updated": 2017-07-01T01:16:43.881Z,
                },
                "created": 2018-01-14T01:36:26.220Z,
                "id": "sv3hmbefpc95MRXQsTHr",
                "sortBy": "",
                "title": "Brave New World",
                "updated": 2018-01-14T01:36:26.220Z,
              }
            }
            disabled={false}
            errors={
              Object {
                "author": "",
                "sortBy": "",
                "title": "",
              }
            }
            onAuthorChange={[Function]}
            onCancel={[Function]}
            onInputChange={[Function]}
            onSubmit={[Function]}
            submitDisabled={false}
            topLevelError=""
          />,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "text": "Update Book",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "authors": Array [
              Object {
                "bookCount": 1,
                "created": 2017-07-01T01:16:43.881Z,
                "firstName": "Aldous",
                "id": "oiw223h3qndlksadffl3ksdf",
                "lastName": "Huxley",
                "updated": 2017-07-01T01:16:43.881Z,
              },
              Object {
                "bookCount": 1,
                "created": 2017-07-06T22:02:11.504Z,
                "firstName": "Kurt",
                "id": "9328fhs938hfdsdblkhf98",
                "lastName": "Vonnegut",
                "updated": 2017-07-06T22:02:11.504Z,
              },
              Object {
                "bookCount": 1,
                "created": 2017-08-17T20:47:35.232Z,
                "firstName": "George R. R.",
                "id": "92nd9ab3kjasegfhasf783fu",
                "lastName": "Martin",
                "updated": 2017-08-17T20:47:35.232Z,
              },
              Object {
                "bookCount": 1,
                "created": 2017-07-06T22:02:17.051Z,
                "firstName": "Tom",
                "id": "98fn98w3ucbe8c7b276v",
                "lastName": "Robbins",
                "updated": 2017-08-17T21:27:35.885Z,
              },
              Object {
                "bookCount": 2,
                "created": 2017-07-06T22:02:17.051Z,
                "firstName": "Stephen",
                "id": "93ucbe8c8fn98w7b276v",
                "lastName": "King",
                "updated": 2017-08-17T21:27:35.885Z,
              },
            ],
            "book": Object {
              "author": Object {
                "bookCount": 1,
                "created": 2017-07-01T01:16:43.881Z,
                "firstName": "Aldous",
                "id": "oiw223h3qndlksadffl3ksdf",
                "lastName": "Huxley",
                "updated": 2017-07-01T01:16:43.881Z,
              },
              "created": 2018-01-14T01:36:26.220Z,
              "id": "sv3hmbefpc95MRXQsTHr",
              "sortBy": "",
              "title": "Brave New World",
              "updated": 2018-01-14T01:36:26.220Z,
            },
            "disabled": false,
            "errors": Object {
              "author": "",
              "sortBy": "",
              "title": "",
            },
            "onAuthorChange": [Function],
            "onCancel": [Function],
            "onInputChange": [Function],
            "onSubmit": [Function],
            "submitDisabled": false,
            "topLevelError": "",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": [Function],
    },
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": <Card>
          <CardHeader
            text="Update Book"
          />
          <BookForm
            authors={
              Array [
                Object {
                  "bookCount": 1,
                  "created": 2017-07-01T01:16:43.881Z,
                  "firstName": "Aldous",
                  "id": "oiw223h3qndlksadffl3ksdf",
                  "lastName": "Huxley",
                  "updated": 2017-07-01T01:16:43.881Z,
                },
                Object {
                  "bookCount": 1,
                  "created": 2017-07-06T22:02:11.504Z,
                  "firstName": "Kurt",
                  "id": "9328fhs938hfdsdblkhf98",
                  "lastName": "Vonnegut",
                  "updated": 2017-07-06T22:02:11.504Z,
                },
                Object {
                  "bookCount": 1,
                  "created": 2017-08-17T20:47:35.232Z,
                  "firstName": "George R. R.",
                  "id": "92nd9ab3kjasegfhasf783fu",
                  "lastName": "Martin",
                  "updated": 2017-08-17T20:47:35.232Z,
                },
                Object {
                  "bookCount": 1,
                  "created": 2017-07-06T22:02:17.051Z,
                  "firstName": "Tom",
                  "id": "98fn98w3ucbe8c7b276v",
                  "lastName": "Robbins",
                  "updated": 2017-08-17T21:27:35.885Z,
                },
                Object {
                  "bookCount": 2,
                  "created": 2017-07-06T22:02:17.051Z,
                  "firstName": "Stephen",
                  "id": "93ucbe8c8fn98w7b276v",
                  "lastName": "King",
                  "updated": 2017-08-17T21:27:35.885Z,
                },
              ]
            }
            book={
              Object {
                "author": Object {
                  "bookCount": 1,
                  "created": 2017-07-01T01:16:43.881Z,
                  "firstName": "Aldous",
                  "id": "oiw223h3qndlksadffl3ksdf",
                  "lastName": "Huxley",
                  "updated": 2017-07-01T01:16:43.881Z,
                },
                "created": 2018-01-14T01:36:26.220Z,
                "id": "sv3hmbefpc95MRXQsTHr",
                "sortBy": "",
                "title": "Brave New World",
                "updated": 2018-01-14T01:36:26.220Z,
              }
            }
            disabled={false}
            errors={
              Object {
                "author": "",
                "sortBy": "",
                "title": "",
              }
            }
            onAuthorChange={[Function]}
            onCancel={[Function]}
            onInputChange={[Function]}
            onSubmit={[Function]}
            submitDisabled={false}
            topLevelError=""
          />
        </Card>,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": Array [
            <CardHeader
              text="Update Book"
            />,
            <BookForm
              authors={
                Array [
                  Object {
                    "bookCount": 1,
                    "created": 2017-07-01T01:16:43.881Z,
                    "firstName": "Aldous",
                    "id": "oiw223h3qndlksadffl3ksdf",
                    "lastName": "Huxley",
                    "updated": 2017-07-01T01:16:43.881Z,
                  },
                  Object {
                    "bookCount": 1,
                    "created": 2017-07-06T22:02:11.504Z,
                    "firstName": "Kurt",
                    "id": "9328fhs938hfdsdblkhf98",
                    "lastName": "Vonnegut",
                    "updated": 2017-07-06T22:02:11.504Z,
                  },
                  Object {
                    "bookCount": 1,
                    "created": 2017-08-17T20:47:35.232Z,
                    "firstName": "George R. R.",
                    "id": "92nd9ab3kjasegfhasf783fu",
                    "lastName": "Martin",
                    "updated": 2017-08-17T20:47:35.232Z,
                  },
                  Object {
                    "bookCount": 1,
                    "created": 2017-07-06T22:02:17.051Z,
                    "firstName": "Tom",
                    "id": "98fn98w3ucbe8c7b276v",
                    "lastName": "Robbins",
                    "updated": 2017-08-17T21:27:35.885Z,
                  },
                  Object {
                    "bookCount": 2,
                    "created": 2017-07-06T22:02:17.051Z,
                    "firstName": "Stephen",
                    "id": "93ucbe8c8fn98w7b276v",
                    "lastName": "King",
                    "updated": 2017-08-17T21:27:35.885Z,
                  },
                ]
              }
              book={
                Object {
                  "author": Object {
                    "bookCount": 1,
                    "created": 2017-07-01T01:16:43.881Z,
                    "firstName": "Aldous",
                    "id": "oiw223h3qndlksadffl3ksdf",
                    "lastName": "Huxley",
                    "updated": 2017-07-01T01:16:43.881Z,
                  },
                  "created": 2018-01-14T01:36:26.220Z,
                  "id": "sv3hmbefpc95MRXQsTHr",
                  "sortBy": "",
                  "title": "Brave New World",
                  "updated": 2018-01-14T01:36:26.220Z,
                }
              }
              disabled={false}
              errors={
                Object {
                  "author": "",
                  "sortBy": "",
                  "title": "",
                }
              }
              onAuthorChange={[Function]}
              onCancel={[Function]}
              onInputChange={[Function]}
              onSubmit={[Function]}
              submitDisabled={false}
              topLevelError=""
            />,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "text": "Update Book",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "authors": Array [
                Object {
                  "bookCount": 1,
                  "created": 2017-07-01T01:16:43.881Z,
                  "firstName": "Aldous",
                  "id": "oiw223h3qndlksadffl3ksdf",
                  "lastName": "Huxley",
                  "updated": 2017-07-01T01:16:43.881Z,
                },
                Object {
                  "bookCount": 1,
                  "created": 2017-07-06T22:02:11.504Z,
                  "firstName": "Kurt",
                  "id": "9328fhs938hfdsdblkhf98",
                  "lastName": "Vonnegut",
                  "updated": 2017-07-06T22:02:11.504Z,
                },
                Object {
                  "bookCount": 1,
                  "created": 2017-08-17T20:47:35.232Z,
                  "firstName": "George R. R.",
                  "id": "92nd9ab3kjasegfhasf783fu",
                  "lastName": "Martin",
                  "updated": 2017-08-17T20:47:35.232Z,
                },
                Object {
                  "bookCount": 1,
                  "created": 2017-07-06T22:02:17.051Z,
                  "firstName": "Tom",
                  "id": "98fn98w3ucbe8c7b276v",
                  "lastName": "Robbins",
                  "updated": 2017-08-17T21:27:35.885Z,
                },
                Object {
                  "bookCount": 2,
                  "created": 2017-07-06T22:02:17.051Z,
                  "firstName": "Stephen",
                  "id": "93ucbe8c8fn98w7b276v",
                  "lastName": "King",
                  "updated": 2017-08-17T21:27:35.885Z,
                },
              ],
              "book": Object {
                "author": Object {
                  "bookCount": 1,
                  "created": 2017-07-01T01:16:43.881Z,
                  "firstName": "Aldous",
                  "id": "oiw223h3qndlksadffl3ksdf",
                  "lastName": "Huxley",
                  "updated": 2017-07-01T01:16:43.881Z,
                },
                "created": 2018-01-14T01:36:26.220Z,
                "id": "sv3hmbefpc95MRXQsTHr",
                "sortBy": "",
                "title": "Brave New World",
                "updated": 2018-01-14T01:36:26.220Z,
              },
              "disabled": false,
              "errors": Object {
                "author": "",
                "sortBy": "",
                "title": "",
              },
              "onAuthorChange": [Function],
              "onCancel": [Function],
              "onInputChange": [Function],
              "onSubmit": [Function],
              "submitDisabled": false,
              "topLevelError": "",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": [Function],
      },
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;
